<a href="https://colab.research.google.com/github/SaharaAli16/UNT_Teaching/blob/main/DTSC_4050/Week3_statlearn_lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Python

<a target="_blank" href="https://colab.research.google.com/github/intro-stat-learning/ISLP_labs/blob/v2.2.1/Ch02-statlearn-lab.ipynb">
<img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>






## Getting Started

To run the labs in the ISLP book, you will need:




 * Google's `Colaboratory` service: [colab.research.google.com/](https://colab.research.google.com/).

Please see the `Python` resources page on the book website [statlearning.com](https://www.statlearning.com) for up-to-date information about getting `Python` and `Jupyter` working on your computer.

You will need to install the `ISLP` package, which provides access to the datasets and custom-built functions that we provide.
The `Python` resources page has a link to the `ISLP` documentation website.




In [ ]:
!pip install ISLP

## Basic Commands


In this lab, we will introduce some simple `Python` commands.
 For more resources about `Python` in general, readers may want to consult the tutorial at [docs.python.org/3/tutorial/](https://docs.python.org/3/tutorial/).





Like most programming languages, `Python` uses *functions*
to perform operations.   To run a
function called `fun`, we type
`fun(input1,input2)`, where the inputs (or *arguments*)
`input1` and `input2` tell
`Python` how to run the function.  A function can have any number of
inputs. For example, the
`print()`  function outputs a text representation of all of its arguments to the console.

In [ ]:
print('fit a model with', 11, 'variables')


fit a model with 11 variables


Adding two integers in `Python` is pretty intuitive.

In [ ]:
3 + 5


8

In `Python`, textual data is handled using
*strings*. For instance, `"hello"` and
`'hello'`
are strings.
We can concatenate them using the addition `+` symbol.

In [ ]:
"hello" + " " + "world"


'hello world'

 A string is actually a type of *sequence*: this is a generic term for an ordered list.
 The three most important types of sequences are lists, tuples, and strings.  
We introduce lists now.

The following command instructs `Python` to join together
the numbers 3, 4, and 5, and to save them as a
*list* named `x`. When we
type `x`, it gives us back the list.

In [ ]:
x = [3, 4, 5]
x


[3, 4, 5]

Note that we used the brackets
`[]` to construct this list.

We will often want to add two sets of numbers together. It is reasonable to try the following code,
though it will not produce the desired results.

In [ ]:
y = [4, 9, 7]
x + y


[3, 4, 5, 4, 9, 7]

The result may appear slightly counterintuitive: why did `Python` not add the entries of the lists
element-by-element?
 In `Python`, lists hold *arbitrary* objects, and  are added using  *concatenation*.
 In fact, concatenation is the behavior that we saw earlier when we entered `"hello" + " " + "world"`.


This example reflects the fact that
 `Python` is a general-purpose programming language. Much of `Python`'s  data-specific
functionality comes from other packages, notably `numpy`
and `pandas`.
In the next section, we will introduce the  `numpy` package.
See [docs.scipy.org/doc/numpy/user/quickstart.html](https://docs.scipy.org/doc/numpy/user/quickstart.html) for more information about `numpy`.


## Introduction to Numerical Python

As mentioned earlier, this book makes use of functionality   that is contained in the `numpy`
 *library*, or *package*. A package is a collection of modules that are not necessarily included in
 the base `Python` distribution. The name `numpy` is an abbreviation for *numerical Python*.

  To access `numpy`, we must first `import` it.

In [ ]:
import numpy as np

In the previous line, we named the `numpy` *module* `np`; an abbreviation for easier referencing.

In `numpy`, an *array* is  a generic term for a multidimensional
set of numbers.
We use the `np.array()` function to define   `x` and `y`, which are one-dimensional arrays, i.e. vectors.

In [ ]:
x = np.array([3, 4, 5])
y = np.array([4, 9, 7])

Note that if you forgot to run the `import numpy as np` command earlier, then
you will encounter an error in calling the `np.array()` function in the previous line.
 The syntax `np.array()` indicates that the function being called
is part of the `numpy` package, which we have abbreviated as `np`.

Since `x` and `y` have been defined using `np.array()`, we get a sensible result when we add them together. Compare this to our results in the previous section,
 when we tried to add two lists without using `numpy`.

In [ ]:
x + y

array([ 7, 13, 12])

In `numpy`, matrices are typically represented as two-dimensional arrays, and vectors as one-dimensional arrays. {While it is also possible to create matrices using  `np.matrix()`, we will use `np.array()` throughout the labs in this book.}
We can create a two-dimensional array as follows.

In [ ]:
x = np.array([[1, 2], [3, 4]])
x

array([[1, 2],
       [3, 4]])

The object `x` has several
*attributes*, or associated objects. To access an attribute of `x`, we type `x.attribute`, where we replace `attribute`
with the name of the attribute.
For instance, we can access the `ndim` attribute of  `x` as follows.

In [ ]:
x.ndim

2

The output indicates that `x` is a two-dimensional array.  
Similarly, `x.dtype` is the *data type* attribute of the object `x`. This indicates that `x` is
comprised of 64-bit integers:

In [ ]:
x.dtype

dtype('int64')

Why is `x` comprised of integers? This is because we created `x` by passing in exclusively integers to the `np.array()` function.
  If
we had passed in any decimals, then we would have obtained an array of
*floating point numbers* (i.e. real-valued numbers).

In [ ]:
np.array([[1, 2], [3.0, 4]]).dtype


dtype('float64')

This documentation indicates that we could create a floating point array by passing a `dtype` argument into `np.array()`.

In [ ]:
np.array([[1, 2], [3, 4]], float).dtype


dtype('float64')

The array `x` is two-dimensional. We can find out the number of rows and columns by looking
at its `shape` attribute.

In [ ]:
x.shape


(2, 2)

A *method* is a function that is associated with an
object.
For instance, given an array `x`, the expression
`x.sum()` sums all of its elements, using the `sum()`
method for arrays.
The call `x.sum()` automatically provides `x` as the
first argument to its `sum()` method.

In [ ]:
x = np.array([1, 2, 3, 4])
x.sum()

10

We could also sum the elements of `x` by passing in `x` as an argument to the `np.sum()` function.

In [ ]:
x = np.array([1, 2, 3, 4])
np.sum(x)

10

 As another example, the
`reshape()` method returns a new array with the same elements as
`x`, but a different shape.
 We do this by passing in a `tuple` in our call to
 `reshape()`, in this case `(2, 3)`.  This tuple specifies that we would like to create a two-dimensional array with
$2$ rows and $3$ columns. {Like lists, tuples represent a sequence of objects. Why do we need more than one way to create a sequence? There are a few differences between tuples and lists, but perhaps the most important is that elements of a tuple cannot be modified, whereas elements of a list can be.}

In what follows, the
`\n` character creates a *new line*.

In [ ]:
x = np.array([1, 2, 3, 4, 5, 6])
print('beginning x:\n', x)
x_reshape = x.reshape((2, 3))
print('reshaped x:\n', x_reshape)


beginning x:
 [1 2 3 4 5 6]
reshaped x:
 [[1 2 3]
 [4 5 6]]


The previous output reveals that `numpy` arrays are specified as a sequence
of *rows*. This is  called *row-major ordering*, as opposed to *column-major ordering*.

`Python` (and hence `numpy`) uses 0-based
indexing. This means that to access the top left element of `x_reshape`,
we type in `x_reshape[0,0]`.

In [ ]:
x_reshape[0, 0]

1

Similarly, `x_reshape[1,2]` yields the element in the second row and the third column
of `x_reshape`.

In [ ]:
x_reshape[1, 2]

6

Similarly, `x[2]` yields the
third entry of `x`.

Now, let's modify the top left element of `x_reshape`.  To our surprise, we discover that the first element of `x` has been modified as well!



In [ ]:
print('x before we modify x_reshape:\n', x)
print('x_reshape before we modify x_reshape:\n', x_reshape)
x_reshape[0, 0] = 5
print('x_reshape after we modify its top left element:\n', x_reshape)
print('x after we modify top left element of x_reshape:\n', x)


x before we modify x_reshape:
 [1 2 3 4 5 6]
x_reshape before we modify x_reshape:
 [[1 2 3]
 [4 5 6]]
x_reshape after we modify its top left element:
 [[5 2 3]
 [4 5 6]]
x after we modify top left element of x_reshape:
 [5 2 3 4 5 6]


Modifying `x_reshape` also modified `x` because the two objects occupy the same space in memory.


    

We just saw that we can modify an element of an array. Can we also modify a tuple? It turns out that we cannot --- and trying to do so introduces
an *exception*, or error.

In [ ]:
my_tuple = (3, 4, 5)
my_tuple[0] = 2


TypeError: 'tuple' object does not support item assignment

We now briefly mention some attributes of arrays that will come in handy. An array's `shape` attribute contains its dimension; this is always a tuple.
The  `ndim` attribute yields the number of dimensions, and `T` provides its transpose.

In [ ]:
x_reshape.shape, x_reshape.ndim, x_reshape.T


((2, 3),
 2,
 array([[5, 4],
        [2, 5],
        [3, 6]]))

Notice that the three individual outputs `(2,3)`, `2`, and `array([[5, 4],[2, 5], [3,6]])` are themselves output as a tuple.

We will often want to apply functions to arrays.
For instance, we can compute the
square root of the entries using the `np.sqrt()` function:

In [ ]:
np.sqrt(x)


array([2.23606798, 1.41421356, 1.73205081, 2.        , 2.23606798,
       2.44948974])

We can also square the elements:

In [ ]:
x**2


array([25,  4,  9, 16, 25, 36])

We can compute the square roots using the same notation, raising to the power of $1/2$ instead of 2.

In [ ]:
x**0.5


array([2.23606798, 1.41421356, 1.73205081, 2.        , 2.23606798,
       2.44948974])

Throughout this book, we will often want to generate random data.
The `np.random.normal()`  function generates a vector of random
normal variables. We can learn more about this function by looking at the help page, via a call to `np.random.normal?`.
The first line of the help page  reads `normal(loc=0.0, scale=1.0, size=None)`.
 This  *signature* line tells us that the function's arguments are  `loc`, `scale`, and `size`. These are *keyword* arguments, which means that when they are passed into
 the function, they can be referred to by name (in any order). {`Python` also uses *positional* arguments. Positional arguments do not need to use a keyword. To see an example, type in `np.sum?`. We see that `a` is a positional argument, i.e. this function assumes that the first unnamed argument that it receives is the array to be summed. By contrast, `axis` and `dtype` are keyword arguments: the position in which these arguments are entered into `np.sum()` does not matter.}
 By default, this function will generate random normal variable(s) with mean (`loc`) $0$ and standard deviation (`scale`) $1$; furthermore,
 a single random variable will be generated unless the argument to `size` is changed.

We now generate 50 independent random variables from a $N(0,1)$ distribution.

In [ ]:
x = np.random.normal(size=50)
x


array([ 0.48217252, -0.67234321, -0.85805608, -0.33039883, -1.62467965,
       -1.15004267, -0.32432107,  0.55548296,  1.27818895, -2.15853856,
        1.26917983,  1.34530026,  0.73463507,  0.40715898, -0.5415699 ,
        2.17612877, -0.11173498, -0.77850206, -0.3715974 ,  0.30673243,
       -1.12091252,  0.84512024, -0.43704279,  0.51318223, -0.04733903,
        0.39320792,  0.06920337, -0.43649396, -0.12888378, -0.81145328,
       -0.25509028,  0.81768177,  1.14109612, -0.87887897,  0.06043576,
        1.34023829,  1.21071042,  0.5301245 ,  0.82374567,  0.0242141 ,
        0.05694048, -0.00690879, -0.41437789,  0.93320153, -0.59911954,
        0.43164549, -0.6468398 , -1.35109935, -1.30370501, -0.12339249])

We create an array `y` by adding an independent $N(50,1)$ random variable to each element of `x`.

In [ ]:
y = x + np.random.normal(loc=50, scale=1, size=50)

The `np.corrcoef()` function computes the correlation matrix between `x` and `y`. The off-diagonal elements give the
correlation between `x` and `y`.

In [ ]:
np.corrcoef(x, y)

array([[1.        , 0.60657693],
       [0.60657693, 1.        ]])

If you're following along in your own `Jupyter` notebook, then you probably noticed that you got a different set of results when you ran the past few
commands. In particular,
 each
time we call `np.random.normal()`, we will get a different answer, as shown in the following example.

In [ ]:
print(np.random.normal(scale=5, size=2))
print(np.random.normal(scale=5, size=2))


[7.25777057 4.70296586]
[-5.05792246  6.96787243]


In order to ensure that our code provides exactly the same results
each time it is run, we can set a *random seed*
using the
`np.random.default_rng()` function.
This function takes an arbitrary, user-specified integer argument. If we set a random seed before
generating random data, then re-running our code will yield the same results. The
object `rng` has essentially all the random number generating methods found in `np.random`. Hence, to
generate normal data we use `rng.normal()`.

In [ ]:
rng = np.random.default_rng(1303)
print(rng.normal(scale=5, size=2))
rng2 = np.random.default_rng(1303)
print(rng2.normal(scale=5, size=2))

[ 4.09482632 -1.07485605]
[ 4.09482632 -1.07485605]


Throughout the labs in this book, we use `np.random.default_rng()`  whenever we
perform calculations involving random quantities within `numpy`.  In principle, this
should enable the reader to exactly reproduce the stated results. However, as new versions of `numpy` become available, it is possible
that some small discrepancies may occur between the output
in the labs and the output
from `numpy`.

The `np.mean()`,  `np.var()`, and `np.std()`  functions can be used
to compute the mean, variance, and standard deviation of arrays.  These functions are also
available as methods on the arrays.

In [ ]:
rng = np.random.default_rng(3)
y = rng.standard_normal(10)
np.mean(y), y.mean()

(-0.1126795190952861, -0.1126795190952861)

In [ ]:
np.var(y), y.var(), np.mean((y - y.mean())**2)

(2.7243406406465125, 2.7243406406465125, 2.7243406406465125)

Notice that by default `np.var()` divides by the sample size $n$ rather
than $n-1$; see the `ddof` argument in `np.var?`.


In [ ]:
np.sqrt(np.var(y)), np.std(y)

(1.6505576756498128, 1.6505576756498128)

The `np.mean()`,  `np.var()`, and `np.std()` functions can also be applied to the rows and columns of a matrix.
To see this, we construct a $10 \times 3$ matrix of $N(0,1)$ random variables, and consider computing its row sums.

In [ ]:
X = rng.standard_normal((10, 3))
X

array([[ 0.22578661, -0.35263079, -0.28128742],
       [-0.66804635, -1.05515055, -0.39080098],
       [ 0.48194539, -0.23855361,  0.9577587 ],
       [-0.19980213,  0.02425957,  1.54582085],
       [ 0.54510552, -0.50522874, -0.18283897],
       [ 0.54052513,  1.93508803, -0.26962033],
       [-0.24355868,  1.0023136 , -0.88645994],
       [-0.29172023,  0.88253897,  0.58035002],
       [ 0.0915167 ,  0.67010435, -2.82816231],
       [ 1.02130682, -0.95964476, -1.66861984]])

Since arrays are row-major ordered, the first axis, i.e. `axis=0`, refers to its rows. We pass this argument into the `mean()` method for the object `X`.

In [ ]:
X.mean(axis=0)

array([ 0.15030588,  0.14030961, -0.34238602])

The following yields the same result.

In [ ]:
X.mean(0)

array([ 0.15030588,  0.14030961, -0.34238602])

## Indexing Data
To begin, we  create a two-dimensional `numpy` array.

In [ ]:
A = np.array(np.arange(16)).reshape((4, 4))
A


array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

Typing `A[1,2]` retrieves the element corresponding to the second row and third
column. (As usual, `Python` indexes from $0.$)

In [ ]:
A[1,2]


np.int64(6)

The first number after the open-bracket symbol `[`
 refers to the row, and the second number refers to the column.

### Indexing Rows, Columns, and Submatrices
 To select multiple rows at a time, we can pass in a list
  specifying our selection. For instance, `[1,3]` will retrieve the second and fourth rows:

In [ ]:
A[[1,3]]


array([[ 4,  5,  6,  7],
       [12, 13, 14, 15]])

To select the first and third columns, we pass in  `[0,2]` as the second argument in the square brackets.
In this case we need to supply the first argument `:`
which selects all rows.

In [ ]:
A[:,[0,2]]


array([[ 0,  2],
       [ 4,  6],
       [ 8, 10],
       [12, 14]])

Now, suppose that we want to select the submatrix made up of the second and fourth
rows as well as the first and third columns. This is where
indexing gets slightly tricky. It is natural to try  to use lists to retrieve the rows and columns:

In [ ]:
A[[1,3],[0,2]]


array([ 4, 14])

 Oops --- what happened? We got a one-dimensional array of length two identical to

In [ ]:
np.array([A[1,0],A[3,2]])


array([ 4, 14])

 Similarly,  the following code fails to extract the submatrix comprised of the second and fourth rows and the first, third, and fourth columns:

In [ ]:
A[[1,3],[0,2,3]]


IndexError: shape mismatch: indexing arrays could not be broadcast together with shapes (2,) (3,) 

We can see what has gone wrong here. When supplied with two indexing lists, the `numpy` interpretation is that these provide pairs of $i,j$ indices for a series of entries. That is why the pair of lists must have the same length. However, that was not our intent, since we are looking for a submatrix.

One easy way to do this is as follows. We first create a submatrix by subsetting the rows of `A`, and then on the fly we make a further submatrix by subsetting its columns.


In [ ]:
A[[1,3]][:,[0,2]]


array([[ 4,  6],
       [12, 14]])

There are more efficient ways of achieving the same result.

The *convenience function* `np.ix_()` allows us  to extract a submatrix
using lists, by creating an intermediate *mesh* object.

In [ ]:
idx = np.ix_([1,3],[0,2,3])
A[idx]


array([[ 4,  6,  7],
       [12, 14, 15]])

Alternatively, we can subset matrices efficiently using slices.
  
The slice
`1:4:2` captures the second and fourth items of a sequence, while the slice `0:3:2` captures
the first and third items (the third element in a slice sequence is the step size).

In [ ]:
A[1:4:2,0:3:2]


array([[ 4,  6],
       [12, 14]])

Why are we able to retrieve a submatrix directly using slices but not using lists?
It's because they are different `Python` types, and
are treated differently by `numpy`.
Slices can be used to extract objects from arbitrary sequences, such as strings, lists, and tuples, while the use of lists for indexing is more limited.




    



    



### Boolean Indexing
In `numpy`, a *Boolean* is a type  that equals either   `True` or  `False` (also represented as $1$ and $0$, respectively).
The next line creates a vector of $0$'s, represented as Booleans, of length equal to the first dimension of `A`.

In [ ]:
keep_rows = np.zeros(A.shape[0], bool)
keep_rows

array([False, False, False, False])

We now set two of the elements to `True`.

In [ ]:
keep_rows[[1,3]] = True
keep_rows


array([False,  True, False,  True])

Note that the elements of `keep_rows`, when viewed as integers, are the same as the
values of `np.array([0,1,0,1])`. Below, we use  `==` to verify their equality. When
applied to two arrays, the `==`   operation is applied elementwise.

In [ ]:
np.all(keep_rows == np.array([0,1,0,1]))


True

(Here, the function `np.all()` has checked whether
all entries of an array are `True`. A similar function, `np.any()`, can be used to check whether any entries of an array are `True`.)

   However, even though `np.array([0,1,0,1])`  and `keep_rows` are equal according to `==`, they index different sets of rows!
The former retrieves the first, second, first, and second rows of `A`.

In [ ]:
A[np.array([0,1,0,1])]


array([[0, 1, 2, 3],
       [4, 5, 6, 7],
       [0, 1, 2, 3],
       [4, 5, 6, 7]])

 By contrast, `keep_rows` retrieves only the second and fourth rows  of `A` --- i.e. the rows for which the Boolean equals `True`.

In [ ]:
A[keep_rows]


array([[ 4,  5,  6,  7],
       [12, 13, 14, 15]])

This example shows that Booleans and integers are treated differently by `numpy`.

We again make use of the `np.ix_()` function
 to create a mesh containing the second and fourth rows, and the first,  third, and fourth columns. This time, we apply the function to Booleans,
 rather than lists.

In [ ]:
keep_cols = np.zeros(A.shape[1], bool)
keep_cols[[0, 2, 3]] = True
idx_bool = np.ix_(keep_rows, keep_cols)
A[idx_bool]


array([[ 4,  6,  7],
       [12, 14, 15]])

We can also mix a list with an array of Booleans in the arguments to `np.ix_()`:

In [ ]:
idx_mixed = np.ix_([1,3], keep_cols)
A[idx_mixed]


array([[ 4,  6,  7],
       [12, 14, 15]])

For more details on indexing in `numpy`, readers are referred
to the `numpy` tutorial mentioned earlier.


## Loading Data

Data sets often contain different types of data, and may have names associated with the rows or columns.
For these reasons, they typically are best accommodated using a
 *data frame*.
 We can think of a data frame  as a sequence
of arrays of identical length; these are the columns. Entries in the
different arrays can be combined to form a row.
 The `pandas`
library can be used to create and work with data frame objects.

### Reading in a Data Set

The first step of most analyses involves importing a data set into
`Python`.  
 Before attempting to load
a data set, we must make sure that `Python` knows where to find the file containing it.
If the
file is in the same location
as this notebook file, then we are all set.
Otherwise,we need to specify the file path.

We will begin by reading in `Auto.csv`, available on the book website. This is a comma-separated file, and can be read in using `pd.read_csv()`:

In [ ]:
import pandas as pd
Auto = pd.read_csv('/content/Auto.csv')
Auto.describe()


,mpg,cylinders,displacement,weight,acceleration,year,origin
count,397.000000,397.000000,397.000000,397.000000,397.000000,397.000000,397.000000
mean,23.515869,5.458438,193.532746,2970.261965,15.555668,75.994962,1.574307
std,7.825804,1.701577,104.379583,847.904119,2.749995,3.690005,0.802549
min,9.000000,3.000000,68.000000,1613.000000,8.000000,70.000000,1.000000
25%,17.500000,4.000000,104.000000,2223.000000,13.800000,73.000000,1.000000
50%,23.000000,4.000000,146.000000,2800.000000,15.500000,76.000000,1.000000
75%,29.000000,8.000000,262.000000,3609.000000,17.100000,79.000000,2.000000
max,46.600000,8.000000,455.000000,5140.000000,24.800000,82.000000,3.000000


Before loading data into `Python`, it is a good idea to view it using
a text editor or other software, such as Microsoft Excel.



We now take a look at the column of `Auto` corresponding to the variable `horsepower`:

In [ ]:
Auto['horsepower']


,horsepower
0,130
1,165
2,150
3,150
4,140
...,...
392,86
393,52
394,84
395,79


We see that the `dtype` of this column is `object`.
It turns out that all values of the `horsepower` column were interpreted as strings when reading
in the data.
We can find out why by looking at the unique values.

In [ ]:
np.unique(Auto['horsepower'])


array(['100', '102', '103', '105', '107', '108', '110', '112', '113',
       '115', '116', '120', '122', '125', '129', '130', '132', '133',
       '135', '137', '138', '139', '140', '142', '145', '148', '149',
       '150', '152', '153', '155', '158', '160', '165', '167', '170',
       '175', '180', '190', '193', '198', '200', '208', '210', '215',
       '220', '225', '230', '46', '48', '49', '52', '53', '54', '58',
       '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70',
       '71', '72', '74', '75', '76', '77', '78', '79', '80', '81', '82',
       '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93',
       '94', '95', '96', '97', '98', '?'], dtype=object)

We see the culprit is the value `?`, which is being used to encode missing values.



To fix the problem, we must provide `pd.read_csv()` with an argument called `na_values`.
Now,  each instance of  `?` in the file is replaced with the
value `np.nan`, which means *not a number*:

In [ ]:
Auto = pd.read_csv('Auto.csv',
                   na_values=['?'])
Auto['horsepower'].sum()


np.float64(40952.0)

The `Auto.shape`  attribute tells us that the data has 397
observations, or rows, and nine variables, or columns.

In [ ]:
Auto.shape


(397, 9)

There are
various ways to deal with  missing data.
In this case, since only five of the rows contain missing
observations,  we choose to use the `Auto.dropna()` method to simply remove these rows.

In [ ]:
Auto_new = Auto.dropna()
Auto_new.shape


(392, 9)

### Basics of Selecting Rows and Columns

We can use `Auto.columns`  to check the variable names.

In [ ]:
Auto = Auto_new # overwrite the previous value
Auto.columns


Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin', 'name'],
      dtype='object')

Accessing the rows and columns of a data frame is similar, but not identical, to accessing the rows and columns of an array.
Recall that the first argument to the `[]` method
is always applied to the rows of the array.  
Similarly,
passing in a slice to the `[]` method creates a data frame whose *rows* are determined by the slice:

In [ ]:
Auto[:3]


,mpg,cylinders,displacement,horsepower,weight,acceleration,year,origin,name
0,18.0,8,307.0,130.0,3504,12.0,70,1,chevrolet chevelle malibu
1,15.0,8,350.0,165.0,3693,11.5,70,1,buick skylark 320
2,18.0,8,318.0,150.0,3436,11.0,70,1,plymouth satellite


Similarly, an array of Booleans can be used to subset the rows:

In [ ]:
idx_80 = Auto['year'] > 80
Auto[idx_80]


,mpg,cylinders,displacement,horsepower,weight,acceleration,year,origin,name
338,27.2,4,135.0,84.0,2490,15.7,81,1,plymouth reliant
339,26.6,4,151.0,84.0,2635,16.4,81,1,buick skylark
340,25.8,4,156.0,92.0,2620,14.4,81,1,dodge aries wagon (sw)
341,23.5,6,173.0,110.0,2725,12.6,81,1,chevrolet citation
342,30.0,4,135.0,84.0,2385,12.9,81,1,plymouth reliant
343,39.1,4,79.0,58.0,1755,16.9,81,3,toyota starlet
344,39.0,4,86.0,64.0,1875,16.4,81,1,plymouth champ
345,35.1,4,81.0,60.0,1760,16.1,81,3,honda civic 1300
346,32.3,4,97.0,67.0,2065,17.8,81,3,subaru
347,37.0,4,85.0,65.0,1975,19.4,81,3,datsun 210 mpg


However, if we pass  in a list of strings to the `[]` method, then we obtain a data frame containing the corresponding set of *columns*.

In [ ]:
Auto[['mpg', 'horsepower']]


,mpg,horsepower
0,18.0,130.0
1,15.0,165.0
2,18.0,150.0
3,16.0,150.0
4,17.0,140.0
...,...,...
392,27.0,86.0
393,44.0,52.0
394,32.0,84.0
395,28.0,79.0


Since we did not specify an *index* column when we loaded our data frame, the rows are labeled using integers
0 to 396.

In [ ]:
Auto.index


Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            387, 388, 389, 390, 391, 392, 393, 394, 395, 396],
           dtype='int64', length=392)

We can use the
`set_index()` method to re-name the rows using the contents of `Auto['name']`.

In [ ]:
Auto_re = Auto.set_index('name')
Auto_re


,mpg,cylinders,displacement,horsepower,weight,acceleration,year,origin
name,,,,,,,,
chevrolet chevelle malibu,18.0,8,307.0,130.0,3504,12.0,70,1
buick skylark 320,15.0,8,350.0,165.0,3693,11.5,70,1
plymouth satellite,18.0,8,318.0,150.0,3436,11.0,70,1
amc rebel sst,16.0,8,304.0,150.0,3433,12.0,70,1
ford torino,17.0,8,302.0,140.0,3449,10.5,70,1
...,...,...,...,...,...,...,...,...
ford mustang gl,27.0,4,140.0,86.0,2790,15.6,82,1
vw pickup,44.0,4,97.0,52.0,2130,24.6,82,2
dodge rampage,32.0,4,135.0,84.0,2295,11.6,82,1


In [ ]:
Auto_re.columns


Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin'],
      dtype='object')

We see that the column `'name'` is no longer there.

Now that the index has been set to `name`, we can  access rows of the data
frame by `name` using the `{loc[]`} method of
`Auto`:

In [ ]:
rows = ['amc rebel sst', 'ford torino']
Auto_re.loc[rows]


,mpg,cylinders,displacement,horsepower,weight,acceleration,year,origin
name,,,,,,,,
amc rebel sst,16.0,8,304.0,150.0,3433,12.0,70,1
ford torino,17.0,8,302.0,140.0,3449,10.5,70,1


As an alternative to using the index name, we could retrieve the 4th and 5th rows of `Auto` using the `{iloc[]`} method:

In [ ]:
Auto_re.iloc[[3,4]]


,mpg,cylinders,displacement,horsepower,weight,acceleration,year,origin
name,,,,,,,,
amc rebel sst,16.0,8,304.0,150.0,3433.0,12.0,70,1
ford torino,17.0,8,302.0,140.0,3449.0,10.5,70,1


We can also use it to retrieve the 1st, 3rd and and 4th columns of `Auto_re`:

In [ ]:
Auto_re.iloc[:,[0,2,3]]


,mpg,displacement,horsepower
name,,,
chevrolet chevelle malibu,18.0,307.0,130.0
buick skylark 320,15.0,350.0,165.0
plymouth satellite,18.0,318.0,150.0
amc rebel sst,16.0,304.0,150.0
ford torino,17.0,302.0,140.0
...,...,...,...
ford mustang gl,27.0,140.0,86.0
vw pickup,44.0,97.0,52.0
dodge rampage,32.0,135.0,84.0


We can extract the 4th and 5th rows, as well as the 1st, 3rd and 4th columns, using
a single call to `iloc[]`:

In [ ]:
Auto_re.iloc[[3,4],[0,2,3]]


,mpg,displacement,horsepower
name,,,
amc rebel sst,16.0,304.0,150.0
ford torino,17.0,302.0,140.0


Index entries need not be unique: there are several cars  in the data frame named `ford galaxie 500`.

In [ ]:
Auto_re.loc['ford galaxie 500', ['mpg', 'origin']]


,mpg,origin
name,,
ford galaxie 500,15.0,1
ford galaxie 500,14.0,1
ford galaxie 500,14.0,1


### More on Selecting Rows and Columns
Suppose now that we want to create a data frame consisting of the  `weight` and `origin`  of the subset of cars with
`year` greater than 80 --- i.e. those built after 1980.
To do this, we first create a Boolean array that indexes the rows.
The `loc[]` method allows for Boolean entries as well as strings:

In [ ]:
idx_80 = Auto_re['year'] > 80
Auto_re.loc[idx_80, ['weight', 'origin']]


,weight,origin
name,,
plymouth reliant,2490.0,1
buick skylark,2635.0,1
dodge aries wagon (sw),2620.0,1
chevrolet citation,2725.0,1
plymouth reliant,2385.0,1
toyota starlet,1755.0,3
plymouth champ,1875.0,1
honda civic 1300,1760.0,3
subaru,2065.0,3


In [ ]:
total = 0
for value in [2,3,19]:
    for weight in [3, 2, 1]:
        total += value * weight
print('Total is: {0}'.format(total))

Total is: 144


Above, we summed over each combination of `value` and `weight`.
We also took advantage of the *increment* notation
in `Python`: the expression `a += b` is equivalent
to `a = a + b`. Besides
being a convenient notation, this can save time in computationally
heavy tasks in which the intermediate value of `a+b` need not
be explicitly created.

Perhaps a more
common task would be to sum over `(value, weight)` pairs. For instance,
to compute the average value of a random variable that takes on
possible values 2, 3 or 19 with probability 0.2, 0.3, 0.5 respectively
we would compute the weighted sum. Tasks such as this
can often be accomplished using the `zip()`  function that
loops over a sequence of tuples.

In [ ]:
total = 0
for value, weight in zip([2,3,19],
                         [0.2,0.3,0.5]):
    total += weight * value
print('Weighted average is: {0}'.format(total))


Weighted average is: 10.8


### String Formatting
In the code chunk above we also printed a string
displaying the total. However, the object `total`
is an  integer and not a string.
Inserting the value of something into
a string is a common task, made
simple using
some of the powerful string formatting
tools in `Python`.
Many data cleaning tasks involve
manipulating and programmatically
producing strings.

For example we may want to loop over the columns of a data frame and
print the percent missing in each column.
Let’s create a data frame `D` with columns in which 20% of the entries are missing i.e. set
to `np.nan`.  We’ll create the
values in `D` from a normal distribution with mean 0 and variance 1 using `rng.standard_normal()`
and then overwrite some random entries using `rng.choice()`.

In [ ]:
rng = np.random.default_rng(1)
A = rng.standard_normal((127, 5))
M = rng.choice([0, np.nan], p=[0.8,0.2], size=A.shape)
A += M
D = pd.DataFrame(A, columns=['food',
                             'bar',
                             'pickle',
                             'snack',
                             'popcorn'])
D[:3]


,food,bar,pickle,snack,popcorn
0,0.345584,0.821618,0.330437,-1.303157,NaN
1,NaN,-0.536953,0.581118,0.364572,0.294132
2,NaN,0.546713,NaN,-0.162910,-0.482119


In [ ]:
for col in D.columns:
    template = 'Column "{0}" has {1:.2%} missing values'
    print(template.format(col,
          np.isnan(D[col]).mean()))


Column "food" has 16.54% missing values
Column "bar" has 25.98% missing values
Column "pickle" has 29.13% missing values
Column "snack" has 21.26% missing values
Column "popcorn" has 22.83% missing values


We see that the `template.format()` method expects two arguments `{0}`
and `{1:.2%}`, and the latter includes some formatting
information. In particular, it specifies that the second argument should be expressed as a percent with two decimal digits.

The reference
[docs.python.org/3/library/string.html](https://docs.python.org/3/library/string.html)
includes many helpful and more complex examples.